aeternity/oracle.py

```diff
@@ -0,0 +1,348 @@
+        logger.debug('sending: %s', message)
+        self.websocket.send(json.dumps(message))
+
+
+class EpochRequestError(Exception):
+    pass
+
+
+class EpochClient:
+    next_block_poll_interval_sec = 10
+
+    def __init__(self, config=None):
+        if config is None:
+            config = Config.get_default()
+        self._config = config
+        self._listeners = defaultdict(list)
+        self._connection = Connection(config=config)
+        self._top_block = None
+        # self.send_and_receive(
+        #     {"target":"chain", "action":"subscribe", "payload":{"type":"mined_block"}},
+        #     print
+        # )
+
+    def http_call(self, method, base_url, endpoint, **kwargs):
+        url = base_url + '/' + endpoint
+        response = requests.request(method, url, **kwargs)
+        if response.status_code >= 500:
+            raise EpochRequestError(response)
+        try:
+            return response.json()
+        except JSONDecodeError:
+            raise EpochRequestError(response.text)
+
+    def internal_http_get(self, endpoint, **kwargs):
+        return self.http_call(
+            'get', self._config.internal_api_url, endpoint, **kwargs
+        )
+
+    def internal_http_post(self, endpoint, **kwargs):
+        return self.http_call(
+            'post', self._config.internal_api_url, endpoint, **kwargs
+        )
+
+    def local_http_get(self, endpoint, **kwargs):
+        return self.http_call(
+            'get', self._config.http_api_url, endpoint, **kwargs
+        )
+
+    def local_http_post(self, endpoint, **kwargs):
+        return self.http_call(
+            'post', self._config.http_api_url, endpoint, **kwargs
+        )
+
+    def get_pubkey(self):
+        return self._config.get_pubkey()
+
+    def get_top_block(self):
+        data = requests.get(self._config.top_block_url).json()
+        return int(data['height'])
+
+    def update_top_block(self):
+        self._top_block = self.get_top_block()
+
+    def wait_for_next_block(self):
+        if self._top_block == None:
+            self.update_top_block()
+        while True:
+            time.sleep(self.next_block_poll_interval_sec)
+            new_block = self.get_top_block()
+            if (new_block > self._top_block):
+                self._top_block = new_block
+                break
+
+    def add_listener(self, target, action, callback):
+        self._listeners[(target, action)].append(callback)
+
+    def remove_listener(self, target, action, callback):
+        self._listeners[(target, action)].remove(callback)
+
+    def dispatch_message(self, message):
+        for callback in self._listeners[(message['origin'], message['action'])]:
+            callback(message)
+
+    def mount(self, component):
+        for target, action, callback_name in component.message_listeners:
+            callback = getattr(component, callback_name)
+            self.add_listener(target, action, callback)
+            component.on_mounted(self)
```

Comment:
Not sure about the indentation of `on_mounted` here

Reply:
Nice catch! Since currently there is only one listener for the Oracle, this didn't matter but this would have been a bug later on. Thanks :)